### PR TITLE
Add a document describing how comm=ofi assures Chapel MCM conformance.

### PR DESCRIPTION
**Pull request by gbtitus**
The ofi comm layer does some modestly complicated things to ensure MCM conformance,
and it has some bugs in doing so.  Document all that, as a guide for near-term development
and a reference for a future hand-off to some other "owner".

While here, fix a couple of very minor typos in the spec's MCM chapter.

### CODE REVIEW

**Comment by gbtitus** [comment]
@mppf You may be interested in this as well.

File: runtime/src/comm/ofi/README.md [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+### Background
+
+The comm layer can use three different kinds of libfabric operations for
+its various transactions.  It uses the RMA (READ/WRITE) interface for
+direct transfers between local and remote memory.  It uses the atomics
+interface for native atomic operations on networks that support those.
+And, it uses the message (SEND/RECV) interface for Active Messages (AMs)
+for executeOn transactions for on-statement bodies, memory transfers
+between unregistered memory, atomic operations which cannot be done
+natively, and several other things.
```
Comment (ronawho) [inline runtime/src/comm/ofi/README.md:35]: Remove "and several other things" or expand on it?

File: runtime/src/comm/ofi/README.md [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+For operations that return data to the initiator, such as RMA GETs and
+native atomic-fetch, the source endpoint is also considered a
+destination endpoint and the default completion level is
+delivery-complete.  The default for other operations is determined by
+the provider.
+
+**_Note (Bug):_** The comm layer assumes that the default completion
+  level for all providers is transmit-complete.  This is a bug.  A
+  provider could certainly choose some other default.
```
Comment (ronawho) [inline runtime/src/comm/ofi/README.md:88]: Can you say more about this? (responding to this comment is fine, I'm just not sure I'm familiar with this bug)

**Review by ronawho — APPROVED** [review]
What's here looks good to me, and definitely helps my understand the ofi implementation better.

My main concern with documents like this (and even long comments in code) is that they will become out of date and then it becomes hard to know what to believe. I think the benefit of this doc outweighs that risk, I'm just wondering if there's something we can do to help us keep it up to date (like an internal issue to remind us to look at this again around release times or something.)

File: runtime/src/comm/ofi/README.md (hunk 1 above)
Comment (ronawho) [reply runtime/src/comm/ofi/README.md:35]: yeah, thanks

File: runtime/src/comm/ofi/README.md (hunk 2 above)
Comment (gbtitus) [reply runtime/src/comm/ofi/README.md:88]: Yeah, after I wrote that I began wondering if it really was a bug.  Providers are free to choose any default completion level they want for operations that don't deliver a local result, such as writes, which is what this is talking about.  As discussed in the paragraph just preceding this, the required completion level for operations that do deliver a local result, such as reads, is delivery-complete.  I think actually that the things the comm layer does to ensure MCM compliance in the absence of delivery-complete for writes and such is sufficient no matter what the provider's default is.  So what I should do is to rephrase this section to say that the comm layer can work with either delivery-complete or whatever the provider's default is, if it's less.  I'll do that.

**Comment by gbtitus** [comment]
> My main concern with documents like this [...] is that they will become out of date and then it becomes hard to know what to believe.

Yeah, definitely a concern.  I think the doc can have ongoing value for developers, but obviously only if it's maintained along with the code.  I could salt the code in the appropriate places with notes saying something like "Update the README if this changes." or, perhaps better, "See README.md in this directory for more information."

I'm not sure an Issue will help.  We could just as easily forget to look at that as we could forget to update the README when the code changes.

**Review by ronawho — APPROVED** [review]

**Pull request merged**